dev/principles.md

```diff
@@ -0,0 +1,1281 @@
+
+Or,
+
+ - `When renaming (removing) an entity, please search the codebase for textual references manually. See
+ https://github.com/apache/incubator-druid/tree/master/dev/principles.md#search-rename`
+
+<a href="lead-yourself"/>
+<h3><a href="#lead-yourself">#</a> 3.7. Always follow these principles no matter if other people do that</h3>
+
+Use these principles in all Druid development and communication regardless of whether the other parties (reviewers,
+authors, discussion participants) also follow the principles or not. Set an example and lead yourself. [Respect people's
+right not to follow these principles](#respect-principles-ignoring).
+
+<h4>3.7.a. Don't rely on other people to put the principles into practice</h4>
+
+The principles are redundant: even if only one participant in an issue or PR (either author or reviewer) the outcome
+should be the same as if all participants followed the principles. Don't count on this: always be that participant who
+asks the [conceptual](#review-problem-and-solution) and [design](#design) questions, [checks that the PR is
+reviewed](#ensure-review) on [different levels](#review-levels), [features are documented
+comprehensively](#document-comprehensively), etc.
+
+<h2>4. Operational principles</h2>
+
+Operational principles aim at minimizing the overhead of highly asynchronous collaboration in issues and PRs.
+
+<h3>4.1. Open issues before doing a lot of work</h3>
+
+Create an issue before starting work on something (code or design) that you think will take you substantial time to
+complete. Briefly describe the problem and the solution that you are planning to work out (or, at least, a direction in
+which you are looking for a solution). There are two reasons for doing this:
+
+ - Signal the fact that you are working on a problem and avoid a potential clash with someone else.
+ - Other developers may chime in with ideas and questions early on which may affect the direction of your work and save
+ you time.
+
+If you are working on a design rather than code, you can later post your results in the same issue that you opened
+before instead of creating another separate "proposal" issue.
+
+<a name="describe-ideas"/>
+<h4><a href="#describe-ideas">#</a> 4.1.a. Describe ideas in issues even if you are not planning to work on them</h4>
+
+Whenever you have an idea (even a vague one) about any new feature in Druid or how the project or a particular place in
+the codebase can be improved, don't hesitate to open an issue (or add a comment to an existing issue if there is one
+relevant). Ideas *shouldn't* be "thought out well" to be worth sharing. See also principle [Open more issues
+yourself](#open-more-issues).
+
+<a name="small-pr">
+<h3><a href="#small-pr">#</a> 4.2. Split your work into reasonably-sized PRs</h3>
+
+<h4>4.2.a. While working on a large project, think early about how you are going to spit you work in reasonably-sized
```

Comment:
I know what you are hinting here, but I don't think it possible to do too small PR's. For me, PR should be an atomic piece of work. For example, when a new feature or fix doesn't work, it can easily be reverted by removing a commit.

Reply:
I don't understand what you mean by this comment. That it's not physically possible to make "too small" PRs? Also don't you intend to leave this comment to principle 4.3?

Reply:
I prefer PR's to be a bit too small than too big. For me, this makes reviewing much easier. I would leave out the section on small PR's. Instead of stating _Reasonable sized PR's_, I'd rather go for _Manageable PR's_.